### PR TITLE
[WIP] High level api

### PR DESCRIPTION
**Pull request by HugoGranstrom**
Hi, I found this project and it seemed like a good first project for interfacing with C (all the hard work has already been done by you 😉). Before I went any further with this I wanted to hear from you if my approach is reasonable or if it has any big flaws. This is my first time working with pointers so I may have gotten something wrong. I wrote a few tests that can be run with `nimble test` and they seem to work. 

My thought is that we call the `ptr mpd_t` `DecimalType` instead for readablitity and we can initialize from ints and strings with `newDecimal()`. 

What do you think? 😄

### CODE REVIEW

File: decimal/decimal.nim [hunk 1]
```diff
@@ -6,3 +6,93 @@
...
+    DecimalType = ptr mpd_t
+
+const
+    DEFAULT_PREC = MPD_RDIGITS * 2
+    DEFAULT_EMAX = when (sizeof(int) == 8): 999999999999999999 else: 425000000
+    DEFAULT_EMIN = when (sizeof(int) == 8): -999999999999999999 else: -425000000
+
+var CTX: mpd_context_t
+var CTX_ADDR = addr CTX
+mpd_defaultcontext(CTX_ADDR)
```
Comment (mratsim) [inline decimal/decimal.nim:21]: There should be an `addQuitProc` statement that destroys/releases the context (or maybe a `defer`)

I don't find any related proc though :/. If you find one, note that you might have to add {.noconv.} pragma to that release proc otherwise addQuitProc will complain.

The main limitation of that approach of hiding the context is that for multi-threading you need one context per thread, but it's fine to start with single-threaded support and redirect people who need multi-threading to the low-level API.

Alternatively there is a `getContext` proc that checks if the current thread as a context already allocated, if yes retrieves it if no create one.

For that CTX should be a {.threadvar.}, you can see an example of this here: https://github.com/status-im/nim-eth-keys/blob/b09f5ee894183d8e49c02b0d9c21c617a7d1ce6b/eth_keys/libsecp256k1.nim#L50-L119

File: decimal/decimal.nim [hunk 2]
```diff
@@ -6,3 +6,93 @@
...
+
+proc setPrec*(prec: int) =
+    if 0 < prec:
+        let success = mpd_qsetprec(CTX_ADDR, prec)
+        if success == 0:
+            echo "Bad prec" 
+
+proc `$`*(s: DecimalType): cstring = mpd_to_sci(s, 0)
+
+proc initDecimal*(): DecimalType = mpd_new(CTX_ADDR)
```
Comment (mratsim) [inline decimal/decimal.nim]: The convention in Nim is `initFoo` for value types and `newFoo` for reference types.
This should be renamed newDecimal.

File: decimal/decimal.nim [hunk 3]
```diff
@@ -6,3 +6,93 @@
 # at your option. This file may not be copied, modified, or distributed except according to those terms.
 
 import decimal_lowlevel
+
+type 
+    DecimalType = ptr mpd_t
```
Comment (mratsim) [inline decimal/decimal.nim]: We should use a ref ptr mpd_t so that it's tracked by the GC and it's not leaked.
For the high-level API people will forgot to release memory.

Example:

Base type:
https://github.com/mratsim/Arraymancer/blob/686836e8dea45fa3838acaf97c757889e1709fa7/src/tensor/data_structure.nim#L58

Auto-alloc/dealloc
https://github.com/mratsim/Arraymancer/blob/686836e8dea45fa3838acaf97c757889e1709fa7/src/tensor/backend/cuda.nim#L23-L40

File: decimal/decimal.nim [hunk 4]
```diff
@@ -6,3 +6,93 @@
...
+proc `/=`*(a, b: DecimalType) =
+    mpd_div(a, a, b, CTX_ADDR)
+
+proc `^`*(a, b: DecimalType): DecimalType =
+    result = initDecimal()
+    mpd_pow(result, a, b, CTX_ADDR)
+
+proc divint*(a, b: DecimalType): DecimalType =
+    result = initDecimal()
+    mpd_divint(result, a, b, CTX_ADDR)
```
Comment (mratsim) [inline decimal/decimal.nim]: It might be worthwhile to use `//` here, but it doesn't seem to be a convention in Nim as we usually overload `/`.

No preference.

**Review by mratsim — CHANGES_REQUESTED** [review]
Nice start, here are a couple of pointers to improve the wrapping

File: decimal/decimal.nim (hunk 1 above)
Comment (HugoGranstrom) [reply decimal/decimal.nim:21]: Regarding the hiding of the context, I couldn't come up with a way to pass a context when doing for example `a + b`. Is there any way of doing that neatly?

File: decimal/decimal.nim (hunk 2 above)
Comment (HugoGranstrom) [reply decimal/decimal.nim]: fixed :-)

File: decimal/decimal.nim (hunk 4 above)
Comment (HugoGranstrom) [reply decimal/decimal.nim]: for now both can be used

File: decimal/decimal.nim (hunk 1 above)
Comment (arnetheduck) [reply decimal/decimal.nim:21]: in 1990 it might have been ok to create a single-threaded api, but it feels a bit off nowadays - except perhaps that nim multi-threading is too broken to use for now ;) `addQuitProc` is pretty useless in general though - if the process is terminating, there's no need to release the context really as the OS will take care of it, and if it's aborting it won't run. 

Destroying contexts is needed mostly if you're doing a proper API that allows you to control the lifetime of the library or use it in multiple parts of your application independently and without leaks (ie plugins or modules that don't know about each other).

threadvar comes with its own problems including the point / time of initialization, and no (easy) way to coordinate release - what happens if you want to pass an decimal between threads?

are mpc contexts re-entrant/thread safe? many contexts are (because they contain only read-only stuff), in that case you don't need on per thread - I'd check the upstream docs here.

File: decimal/decimal.nim (hunk 3 above)
Comment (HugoGranstrom) [reply decimal/decimal.nim]: Have I understood you correctly?: 
We make DecimalType an object and gives it two field, one `Fdata: ptr mpd_t` and one `Fref_tracking: ref[ptr mpd_t]`. `Fdata` is what I called the entire `DecimalType`. `Fref_tracking` is there just to make sure it get GC:ed and it is just an initilized `ref[ptr mpd_t]` and does not point to the same data as `Fdata`.

File: decimal/decimal.nim (hunk 1 above)
Comment (mratsim) [reply decimal/decimal.nim:21]: Multiple threads reading from the same context is supported (https://www.bytereef.org/mpdecimal/doc/libmpdec/functions.html#thread-safety)

The context looks to be read-only: see [specification](http://speleotrove.com/decimal/decarith.html) and the [context part](http://speleotrove.com/decimal/daconcep.html)

This means that we can keep the currnet implementation without jumping through hoops.

File: decimal/decimal.nim (hunk 3 above)
Comment (mratsim) [reply decimal/decimal.nim]: We can even make it a `ref[ptr mpd_t]` directly.

I used a ref and a raw pointer in Arraymancer to avoid double indirection and because CudaTensor are long-lived.

I don't think it applies in our case:
  - Even if there is a double-indirection the addresses will be hot in cache so the cost is low. In Arraymancer you often do a lot of tensor stuff without actually accessing the data.
  - The mp numbers are short-lived and will probably be created/destroyed often, allocating two fields instead of one in that case is probably not justified.

Comment (HugoGranstrom) [reply decimal/decimal.nim]: That sounds really good :-D
So all I have to change basically is to cast all return values from mpd procs to DecimalType. And if `d: DecimalType` I access the pointer with `d[]`?

Comment (HugoGranstrom) [reply decimal/decimal.nim]: And another question: CTX is just an object so it will be GC:ed without having to use `ref`? Or does it matter really, as it lives as long as the program does?

Comment (HugoGranstrom) [reply decimal/decimal.nim]: I solved it like this:
```nim
type DecimalType = ref[ptr mpd_t]

proc newDecimal*(): DecimalType = 
    new[ptr mpd_t](result)
    result[] = mpd_new(CTX_ADDR)

proc newDecimal*(s: string): DecimalType =
    new[ptr mpd_t](result)
    result[] = mpd_new(CTX_ADDR)
    mpd_set_string(result[], s, CTX_ADDR)

proc `+`*(a, b: DecimalType): DecimalType =
    result = newDecimal()
    mpd_add(result[], a[], b[], CTX_ADDR)
```
It works but I'm not quite sure why it does. Shouldn't `result` already be initilized as a `ref[ptr mpd_t]`? And if so, when I run `new` on it, shouldn't I get a ref of a ref? Lot's of questions from a beginner 😅 I'll have to read up on this a bit more I think. The rest of the high level API shouldn't need much more new knowledge though.

Comment (mratsim) [reply decimal/decimal.nim]: This is more idiomatic

```Nim
type DecimalType = ref[ptr mpd_t]

proc newDecimal*(): DecimalType = 
    new result
    result[] = mpd_new(CTX_ADDR)

proc newDecimal*(s: string): DecimalType =
    new result
    result[] = mpd_new(CTX_ADDR)
    mpd_set_string(result[], s, CTX_ADDR)

proc `+`*(a, b: DecimalType): DecimalType =
    result = newDecimal()
    mpd_add(result[], a[], b[], CTX_ADDR)
```

Nim compiler will automatically infer the generic type `ptr mpd_t`.

Here is how it works.
By default Nim initializes types to binary zero, sequences to empty sequences and strings to empty strings.

For ref types, binary zero means a nil pointer.
Also remember that Nim provides an implicit `result` variable.

So just after this:

```Nim
proc newDecimal*(): DecimalType = 
    discard
```

1. You have a `result` of type DecimalType and of value `nil`.
2. You need to initialize it with `new` because the underlying value `ptr mpd_t` needs to be stored somewhere.
3. `new` will do the plumbing: allocate memory, add the GC tracking hooks and initialize the memory location to binary zero as well.
4. Then you access that memory location and assign its value `mpd_new(CTX_ADDR)

In summary, unlike in other language this `new` doesn't create a new instance of an object type. You have your nil object of your desired type, pass it to `new` and it will instantiate it.

There is a [`new` overload that accepts a typedesc](https://nim-lang.org/docs/system.html#new%2Ctypedesc) and can be used as it's common in other language s:

```Nim
proc newDecimal*(): DecimalType = 
    result = new DecimalType
```

but we already have an implicit DecimalType that we now overwrite with new DecimalType, so we do more work.

Comment (HugoGranstrom) [reply decimal/decimal.nim]: Thank you that cleared up a lot of brain knots 😉👍😅 

So `new(result)` would "make more sense" as `init(result)` for the uninitiated.

**Comment by HugoGranstrom** [comment]
Now almost all functions from the "Arithmetic Functions" are wrapped (left out a few odd ones). I also tried to make it nimble installable by renaming and moving some things. On my machine, it works at least. Travis and Appveyor seem to disagree (are they on an older version perhaps?). I did also add templates for adding `DecimalType` with `SomeNumber`, I'm a bit split on that matter though as it kind of encourages not using arbitrary precision :/ Was thinking it would make statements like `let a = x ^ 2` more intuitive by templating it to: `let a = x ^ newDecimal($2)`. Any opinions on this?

This should definitely be ready for a test ride now. I have lots of more tests to write, and documenting all procs with a `## doc` but I want an opinion before I procedes with that.

**Comment by mratsim** [comment]
Whoops, this passed under my radar, flooded with Github updates.

You shouldn't move to the `src/` directory structure in the same PR as it makes review very hard. And while I prefer the `src` structure, it triggers several annoying issues with nim imports and nimble:

- https://github.com/nim-lang/Nim/issues/6700
- https://github.com/nim-lang/RFCs/issues/37
- https://github.com/nim-lang/nimble/issues/653

Don't hesitate to ping me on IRC/Gitter if I forget again.

**Review by mratsim — CHANGES_REQUESTED** [review]
The last commit can be reverted. I can take care of the directory structure. The other commits are good.

**Comment by HugoGranstrom** [comment]
Hehe 😅 I got a bit too ambitious here. Should have realized it made a big mess of it all 🤣 

Thank for handling it :-) Your feed must be bombarded with updates all the time, amazed that you can handle it 😄

I'll keep on working on this then 😉

**Comment by HugoGranstrom** [comment]
I think that's it 😅 Documentation comments are added to all non-obvious procs and next to everything should be covered in the tests. 
If I have missed some very obvious procs: shout! 🤣🙏

**Review by mratsim — COMMENTED** [review]
LGTM, it's just missing the auto deletion but I'll add it myself

**Comment by HugoGranstrom** [comment]
Awesome 😄 

You know those stuff better than I do 👍😉
